man/hfst-proc2.1
.\" DO NOT MODIFY THIS FILE!  It was generated by help2man 1.40.4.
.TH HFST-PROC2 "1" "October 2014" "HFST" "User Commands"
.SH NAME
hfst-proc2 \- =perform matching/lookup on text streams
.SH SYNOPSIS
.B hfst-proc2
[\fIOPTIONS\fR...] \fITOKENIZER\fR
.SH DESCRIPTION
perform matching/lookup on text streams
.SS "Common options:"
.TP
\fB\-h\fR, \fB\-\-help\fR
Print help message
.TP
\fB\-V\fR, \fB\-\-version\fR
Print version info
.TP
\fB\-v\fR, \fB\-\-verbose\fR
Print verbosely while processing
.TP
\fB\-q\fR, \fB\-\-quiet\fR
Only print fatal erros and requested output
.TP
\fB\-s\fR, \fB\-\-silent\fR
Alias of \fB\-\-quiet\fR
.TP
\fB\-n\fR  \fB\-\-newline\fR
Newline as input separator (default is blank line)
.TP
\fB\-x\fR  \fB\-\-xerox\fR
Xerox output (default)
.PP
Use standard streams for input and output (for now).
.SH "REPORTING BUGS"
Report bugs to <hfst\-bugs@helsinki.fi> or directly to our bug tracker at:
<https://sourceforge.net/tracker/?atid=1061990&group_id=224521&func=browse>
.PP
hfst\-proc2 home page:
<https://kitwiki.csc.fi/twiki/bin/view/KitWiki//HfstProc2>
.br
General help using HFST software:
<https://kitwiki.csc.fi/twiki/bin/view/KitWiki//HfstHome>
.SH COPYRIGHT
Copyright \(co 2010 University of Helsinki,
License GPLv3: GNU GPL version 3 <http://gnu.org/licenses/gpl.html>
.br
This is free software: you are free to change and redistribute it.
There is NO WARRANTY, to the extent permitted by law.


NEWS
=================================
 NEWS for hfst-tools and library
=================================

This file contains all noteworthy changes in HFST development between releases.
For full listing of changes see ChangeLog.

Noteworthy changes in 3.8.1
---------------------------

* (the release of version 3.8.0 was delayed, its changes are also listed here)

* changes to flag diacritic handling

  * now hfst-lexc has options --withFlags, --minimizeFlags and --renameFlags to control
    hyperminimization

  * hfst-xfst has corresponding variables --lexc-with-flags, --lexc-minimize-flags and 
    --lexc-rename-flags

  * hfst-xfst has variable xerox-composition (default ON) and hfst-lexc option 
    --xerox-composition (default ON) to control if flags are matched with unknowns and
    identities in composition (as Xerox does)
 
  * hfst-regexp2fst also has option --xerox-composition (default OFF)

  * command twosided flag-diacritics (tfd) added to hfst-xfst

* added	a (temporary) tool called hfst-proc2 for simple tokenization, to become part of
  something else or be renamed in the future

* fixes to pmatch and optimized-lookup functionalities

* improvements to speed issues in optimized-lookup

* fixes to twolc functionalities

* fixed some issues found by scan-build

* improvements to xre parsing

* added option --Werror to hfst-lexc that treats warnings as errors


Noteworthy changes in 3.7.1
---------------------------

* same mappings with different weights are now handled right in 
  parallel replace rules.

* improvements to optimized lookup format


Noteworthy changes in 3.7.0
---------------------------

* better support for weighted parallel replace rules

* allowing empty contexts in replace rules

* bugfixes and improvements to xfst and pmatch tools

* bugfixes to handling of optimized lookup transducers

* support for weighted containment operator 

* tentatively added option and variable 'encode-weights' to regexp
  and xfst parsers to control determinization of weighted transducers,
  the default value being false (encoding weights prevents non-ending
  determinization in some cases)


Noteworthy changes in 3.6.1
---------------------------

* improvements to xfst, lexc, pmatch and lookup tools


Noteworthy changes in 3.6.0
---------------------------

* hfst-lexc uses native hfst functions

* improvements to pmatch tool

* limited support for optimized lookup transducers in hfst-xfst

* better support for Mac and Windows

* replace rule longest match bug fixed


Noteworthy changes in 3.5.1
---------------------------

* improvements and bugfixes to lexc, pmatch and xfst tools

* added an option --enable-all-tools in configure


Noteworthy changes in 3.5.0
---------------------------

* a new tool hfst-xfst for compiling xfst scripts and using xfst commands
  in interactive mode (NOTE: hfst-xfst was earlier a wrapper for foma, so this
  change breaks backwards compatibility)

* wrapper for foma command line tool is renamed to hfst-foma

* changes to the api

  * made HfstTransducer::harmonize() a public member

  * added a boolean argument to function 'set_final_weights' which defines
    whether existing weights are summed with the new weight or overwritten

  * updated cross product and apply boundary mark in replace rules

* improvements to pmatch tool

  * --verbose output gives neater and more fine-grained information

  * some fixes to ?-expansion

  * more checking and better error diagnostics in case of broken tags

  * time individual compilations

  * fixes to leftmost greediness and overlong input

* changes to other tools

  * added support for weights in prolog format, also added an option 
    --prolog to hfst-fst2txt

  * added option --eliminate-flags to hfst-compare

  * added option --harmonize-flags to hfst-regex2fst

  * lexc parser handles epsilon '0' correctly

  * better support for quoted literals and marker '.#.' in xre parser

  * added warning messages about hfst-special symbols of form
    '@_.*_@' to xre parser that are printed in verbose mode


Noteworthy changes in 3.4.6
---------------------------

* Improvements to pmatch functionalities

* Some fixes to twolc, proc and lexc tools

* New operators added to the regexp parser

* Reading in prolog format supported in hfst-txt2fst


Noteworthy changes in 3.4.5
---------------------------

* Several improvements to pmatch tools

* Function calls supported in xre parser

* A bug in replace rules with term complement in context fixed

* Some fixes to the lexc parser

* Now option combination --obey-flags --random works in hfst-fst2strings


Noteworthy changes in 3.4.4
---------------------------

* foma and sfst backends are enabled by default in configure

* Updated to foma version 0.9.17

* New functions added to HfstTransitionGraph, e.g. longest_string_size and
  top_sort

* pmatch tools updated

* Bugs fixed in xre parser and hfst-twolc

* A boolean argument added to binary transducer functions that controls
  whether symbol harmonization is carried out


Noteworthy changes in 3.4.3
---------------------------

* Better support for compilation on Cygwin.

* Tool hfst-prune-alphabet added.

* A function 'hfst_get_exception()' added to the swig interface that
  can be used to handle exceptions on Python 3.

* Bugfixes in hfst-info, hfst-twolc, hfst-regexp2fst and tagger tools.

* Flag diacritic harmonization problem fixed, now identity and
  unknown symbols are not expanded by flag diacritics.


Noteworthy changes in 3.4.2
---------------------------

* Added regexp substitution and left replace epenthesis to regexp parser.

* A bug in including zlib in foma fixed.

* Now ncurses library is used instead if termcap is not found when 
  building hfst-xfst.


Noteworthy changes in 3.4.1
---------------------------

* Updates to regex parser and optimized lookup transducer

* SWIG bindings now work with both Python version 2 and 3. As FILE* conversions

tools/src/hfst-proc2.cc
//! @file hfst-proc2.cc
//!
//! @brief A demo of a replacement for hfst-proc using pmatch
//!
//! @author HFST Team

//  This program is free software: you can redistribute it and/or modify
//  it under the terms of the GNU General Public License as published by
//  the Free Software Foundation, version 3 of the License.
//
//  This program is distributed in the hope that it will be useful,
//  but WITHOUT ANY WARRANTY; without even the implied warranty of
//  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
//  GNU General Public License for more details.
//
//  You should have received a copy of the GNU General Public License
//  along with this program.  If not, see <http://www.gnu.org/licenses/>.

#ifdef HAVE_CONFIG_H
#  include <config.h>
#endif


#include <iostream>
#include <fstream>

#include <vector>
#include <map>
#include <string>

using std::string;
using std::vector;
using std::pair;

#include <cstdio>
#include <cstdlib>
#include <cstring>
#include <getopt.h>
#include <math.h>
#include <errno.h>

#include "hfst-commandline.h"
#include "hfst-program-options.h"
#include "hfst-tool-metadata.h"
#include "implementations/optimized-lookup/pmatch.h"

#include "inc/globals-common.h"
#include "inc/globals-unary.h"

bool blankline_separated = true;
std::string tokenizer_filename;
enum OutputFormat {
    xerox
};
OutputFormat output_format = xerox;

void
print_usage()
{
    // c.f. http://www.gnu.org/prep/standards/standards.html#g_t_002d_002dhelp
    fprintf(message_out, "Usage: %s [OPTIONS...] TOKENIZER\n"
            "perform matching/lookup on text streams\n"
            "\n", program_name);
    print_common_program_options(message_out);
    fprintf(message_out,
            "  -n  --newline          Newline as input separator (default is blank line)\n"
            "  -x  --xerox            Xerox output (default)\n");
    fprintf(message_out, 
            "Use standard streams for input and output (for now).\n"
            "\n"
        );
    
    print_report_bugs();
    fprintf(message_out, "\n");
    print_more_info();
    fprintf(message_out, "\n");
}

void match_and_print(hfst_ol::PmatchContainer & container,
                     std::ostream & outstream,
                     std::string & input_text)
{
    if (input_text.size() > 0 && input_text.at(input_text.size() - 1) == '\n') {
        // Remove final newline
        input_text.erase(input_text.size() -1, 1);
    }
    hfst_ol::LocationVectorVector locations = container.locate(input_text);

    // Output formatting
    if (output_format == xerox) {
        for(hfst_ol::LocationVectorVector::const_iterator it = locations.begin();
            it != locations.end(); ++it) {
            for (hfst_ol::LocationVector::const_iterator loc_it = it->begin();
                 loc_it != it->end(); ++loc_it) {
                outstream << loc_it->input << "\t" << loc_it->output << std::endl;
            }
            std::cerr << endl;
        }
    } else {
        // More formatting options
    }
    outstream << std::endl;
}

int process_input(hfst_ol::PmatchContainer & container,
                  std::ostream & outstream)
{
    std::string input_text;
    char * line = NULL;
    size_t len = 0;
    while (hfst_getline(&line, &len, inputfile) > 0) {
        if (!blankline_separated) {
            // newline separated
            input_text = line;
            match_and_print(container, outstream, input_text);
        } else if (line[0] == '\n') {
            match_and_print(container, outstream, input_text);
            input_text.clear();
        } else {
            input_text.append(line);
        }
        free(line);
        line = NULL;
    }
    
    if (blankline_separated && !input_text.empty()) {
        match_and_print(container, outstream, input_text);
    }
    return EXIT_SUCCESS;
}


int parse_options(int argc, char** argv)
{
    extend_options_getenv(&argc, &argv);
    // use of this function requires options are settable on global scope
    while (true)
    {
        static const struct option long_options[] =
            {
                HFST_GETOPT_COMMON_LONG,
                {"newline", no_argument, 0, 'n'},
                {"xerox", no_argument, 0, 'x'},
                {0,0,0,0}
            };
        int option_index = 0;
        char c = getopt_long(argc, argv, HFST_GETOPT_COMMON_SHORT "nxl",
                             long_options, &option_index);
        if (-1 == c)
        {
            break;
        }


        switch (c)
        {
#include "inc/getopt-cases-common.h"
        case 'n':
            blankline_separated = false;
            break;
        case 'x':
            output_format = xerox;
            break;
#include "inc/getopt-cases-error.h"
        }

        
        
    }

//            if (!inputNamed)
//        {
//            inputfile = stdin;
//            inputfilename = hfst_strdup("<stdin>");
//        }
        
        // no more options, we should now be at the input filename
        if ( (optind + 1) < argc)
        {
            std::cerr << "More than one input file given\n";
            return EXIT_FAILURE;
        }
        else if ( (optind + 1) == argc)
        {
            tokenizer_filename = argv[(optind)];
            return EXIT_CONTINUE;
        }
        else
        {
            std::cerr << "No input file given\n";
            return EXIT_FAILURE;
        }


#include "inc/check-params-common.h"


    
    return EXIT_FAILURE;
}

int main(int argc, char ** argv)
{
    hfst_set_program_name(argv[0], "0.1", "HfstProc2");
    hfst_setlocale();
    int retval = parse_options(argc, argv);
    if (retval != EXIT_CONTINUE) {
        return retval;
    }
    std::ifstream instream(tokenizer_filename.c_str(),
                           std::ifstream::binary);
    if (!instream.good()) {
        std::cerr << "Could not open file " << tokenizer_filename << std::endl;
        return EXIT_FAILURE;
    }
    hfst_ol::PmatchContainer container(instream, verbose);
//     if (outfile != stdout) {
//         std::filebuf fb;
// fb.open(outfilename, std::ios::out);
// std::ostream outstream(&fb);
// return process_input(container, outstream);
// fb.close();
//     } else {
    return process_input(container, std::cout);
}
